linera-core/src/client/mod.rs

```diff
@@ -1301,46 +1301,76 @@ where
+                certificates
+                    .entry(certificate.value().chain_id())
+                    .or_default()
+                    .insert(certificate.value().height(), certificate);
             }
         }
-        for chain_id in other_sender_chains {
-            // Certificates for this chain were omitted from `certificates` because they were
-            // already processed locally. If they were processed in a concurrent task, it is not
-            // guaranteed that their cross-chain messages were already handled.
-            if let Err(error) = self
-                .client
-                .local_node
-                .retry_pending_cross_chain_requests(chain_id)
-                .await
-            {
-                error!("Failed to retry outgoing messages from {chain_id}: {error}");
+        let certificate_count = certificates.values().map(BTreeMap::len).sum::<usize>();
+
+        tracing::info!(
+            "Received {certificate_count} certificates from {validator_count} validator(s)."
+        );
+
+        // We would like to use all chain workers, but we need to keep some of them free, because
+        // handling the certificates can trigger messages to other chains, and putting these in
+        // the inbox requires the recipient chain's worker, too.
+        let chain_worker_limit = (CHAIN_WORKER_LIMIT / 2).max(1);
+
+        // Process the certificates sorted by chain and in ascending order of block height.
+        let stream = stream::iter(certificates.into_values().map(|certificates| {
+            let client = self.clone();
+            async move {
+                for certificate in certificates.into_values() {
+                    let hash = certificate.hash();
+                    let mode = ReceiveCertificateMode::AlreadyChecked;
+                    if let Err(e) = client.receive_certificate_internal(certificate, mode).await {
+                        warn!("Received invalid certificate {hash}: {e}");
+                    }
+                }
+            }
+        }))
+        .buffer_unordered(chain_worker_limit);
+        stream.for_each(future::ready).await;
+
+        // Certificates for these chains were omitted from `certificates` because they were
+        // already processed locally. If they were processed in a concurrent task, it is not
+        // guaranteed that their cross-chain messages were already handled.
+        let stream = stream::iter(other_sender_chains.into_iter().map(|chain_id| {
+            let local_node = self.client.local_node.clone();
+            async move {
+                if let Err(error) = local_node
+                    .retry_pending_cross_chain_requests(chain_id)
+                    .await
+                {
+                    error!("Failed to retry outgoing messages from {chain_id}: {error}");
+                }
             }
+        }))
+        .buffer_unordered(chain_worker_limit);
+        stream.for_each(future::ready).await;
```

Comment:
I'm not sure about it. Firstly, I hope this is only temporary: We should find a way to just `join_all` here, and have the `WorkerState` handle its actors automatically even if there are lots of requests.

Secondly, the extracted logic would just be two lines:

```rust
        let stream = stream::iter(futures).buffer_unordered(limit);
        stream.for_each(future::ready).await;
```

And even those don't apply to all three places where we use `buffer_unordered`: In one place we want to actually `try_collect` instead of `for_each(future::ready)`.